cloud/docker/docker_login.py

```diff
@@ -0,0 +1,246 @@
+        try:
+            self.response = self.client.login(
+                self.username,
+                password=self.password,
+                email=self.email,
+                registry=self.registry,
+                reauth=self.reauth,
+                dockercfg_path=self.dockercfg_path
+            )
+        except DockerAPIError as e:
+            self.module.fail_json(msg="Docker API Error: %s" % e.explanation)
+        except Exception as e:
+            self.module.fail_json(msg="failed to login to the remote registry", error=repr(e))
+
+        # Get status from registry response.
+        if self.response.has_key("Status"):
+            self.log.append(self.response["Status"])
+            if self.response["Status"] == "Login Succeeded":
+                self.changed = True
+        else:
+            self.log.append("Already Authentificated")
+
+        # Update the dockercfg if changed but not failed.
+        if self.has_changed() and not self.module.check_mode:
+            self.update_dockercfg()
+
+    # This is what the underlaying docker-py unfortunately doesn't do (yet).
+    def update_dockercfg(self):
+
+        # Create dockercfg file if it does not exist.
+        if not os.path.exists(self.dockercfg_path):
+            open(self.dockercfg_path, "w")
+            self.log.append("Created new Docker config file at %s" % self.dockercfg_path)
+        else:
+            self.log.append("Updated existing Docker config file at %s" % self.dockercfg_path)
+
+        # Get existing dockercfg into a dict.
+        try:
+            docker_config = json.load(open(self.dockercfg_path, "r"))
+        except ValueError:
+            docker_config = dict()
+        if not docker_config.has_key(self.registry):
+            docker_config[self.registry] = dict()
+        docker_config[self.registry] = dict(
+            auth  = base64.b64encode(self.username + b':' + self.password),
+            email = self.email
+        )
+
+        # Write updated dockercfg to dockercfg file.
+        try:
+            json.dump(docker_config, open(self.dockercfg_path, "w"), indent=4, sort_keys=True)
+        except Exception as e:
+            self.module.fail_json(msg="failed to write auth details to file", error=repr(e))
+
+    # Compatible to docker-py auth.decode_docker_auth()
+    def encode_docker_auth(self, auth):
+        s = base64.b64decode(auth)
+        login, pwd = s.split(b':', 1)
+        return login.decode('ascii'), pwd.decode('ascii')
+
+    def get_msg(self):
+        return ". ".join(self.log)
+
+    def has_changed(self):
+        return self.changed
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            registry        = dict(required=True),
+            username        = dict(required=True),
+            password        = dict(required=True),
+            email           = dict(required=False, default='anonymous@localhost.local'),
```

Comment:
Ah -- from the commit history and comments, I see that you added this in response to the prior review.  I think that may have been a misunderstanding.  If the docker-py api is fine taking None here, we probably want the default to be None and document that the default is None in the documentation.